# Group Assignment 2023-2024
## Getting you started (Part 1)

In [1]:
# import packages
import numpy as np
import pandas as pd
import os

In [2]:
# define directory to look at
data_dir = "./Data/"

# let's have a look
os.listdir(data_dir)

['customers.csv',
 'Data_Description.pdf',
 'geolocation.csv',
 'orders.csv',
 'order_items.csv',
 'order_payments.csv',
 'order_reviews.csv',
 'Picture1.png',
 'products.csv',
 'sellers.csv']

In [3]:
# NOTE: if you are working on a mac, you may have a .DS_Store file in your directory
# this is a hidden file that is created by macOS

# we can remove it by running the following command:
dir_list = [obs for obs in os.listdir(data_dir) if ".DS" not in obs]
dir_list

['customers.csv',
 'Data_Description.pdf',
 'geolocation.csv',
 'orders.csv',
 'order_items.csv',
 'order_payments.csv',
 'order_reviews.csv',
 'Picture1.png',
 'products.csv',
 'sellers.csv']

<img src="./Data/Picture1.png" align="center" width="800"/>

## 1. Orders
This is the main file for the group assignment. As you can see in the ERD, it shares a lot of connections to the other files.

In [4]:
# import orders
orders = pd.read_csv("./Data/orders.csv")

In [5]:
# show first 5 rows
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  a7c556ca59f5d0c765380ddafd09941d   
1  53cdb2fc8bc7dce0b6741e2150273451  9e285031fb5b57b3d1464d8672b8ed22   
2  47770eb9100c2d0c44946d9cf07ec65d  6348d8e73c3d01bb030637817b424016   
3  949d5b44dbf5de918fe9c16f97b45f8a  9b5d68f669f7ed215f789471556529fe   
4  ad21c59c0840e6cb83a9ceb5573f8159  42510cf6f88a1b5a38b09f09867c1f8d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [6]:
# check data type of columns
orders.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

### Date format
As you can see the date variables (`order_purchase_timestamp`, `order_approved_at`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, and `order_estimated_delivery_date`) are still in the string format, so we will have to convert them into the correct **date format**.

In [7]:
# show first 5 rows
orders[['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']].head()

order_purchase_timestamp    order_approved_at order_delivered_carrier_date  \
0      2017-10-02 10:56:33  2017-10-02 11:07:15          2017-10-04 19:55:00   
1      2018-07-24 20:41:37  2018-07-26 03:24:27          2018-07-26 14:31:00   
2      2018-08-08 08:38:49  2018-08-08 08:55:23          2018-08-08 13:50:00   
3      2017-11-18 19:28:06  2017-11-18 19:45:59          2017-11-22 13:39:59   
4      2018-02-13 21:18:39  2018-02-13 22:20:29          2018-02-14 19:46:34   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-10-10 21:25:13           2017-10-18 00:00:00  
1           2018-08-07 15:27:45           2018-08-13 00:00:00  
2           2018-08-17 18:06:29           2018-09-04 00:00:00  
3           2017-12-02 00:28:42           2017-12-15 00:00:00  
4           2018-02-16 18:17:02           2018-02-26 00:00:00

Notice that all date variables are in the `YYYY-MM-DD HH:MM:SS` format. To get an **overview** of the `strftime()` and `strptime()` Format Codes used in `format` parameter go to https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior

In [8]:
# define a list with all columns that we want to convert to the datetime data type
date_cols = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

# let's convert them to datetime format
for col in date_cols:
    orders[col] = pd.to_datetime(orders[col], format='%Y-%m-%d %H:%M:%S')

# check first five rows
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  a7c556ca59f5d0c765380ddafd09941d   
1  53cdb2fc8bc7dce0b6741e2150273451  9e285031fb5b57b3d1464d8672b8ed22   
2  47770eb9100c2d0c44946d9cf07ec65d  6348d8e73c3d01bb030637817b424016   
3  949d5b44dbf5de918fe9c16f97b45f8a  9b5d68f669f7ed215f789471556529fe   
4  ad21c59c0840e6cb83a9ceb5573f8159  42510cf6f88a1b5a38b09f09867c1f8d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

**NOTE**: the column `order_estimated_delivery_date` only displays the date, because this column only specifies the day of delivery, not the hour. However, Python will deal with this by using 00:00:00 for the time.

In [9]:
# check updated data types
orders.dtypes

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

### Missing values
Let's now investigate whether we have to deal with **missing values**

In [10]:
# check number of missing values per variable in orders dataset
for col in orders.columns:
    missings = len(orders[col][orders[col].isnull()]) / float(len(orders))
    print(col, missings)

order_id 0.0
customer_id 0.0
order_status 0.0
order_purchase_timestamp 0.0
order_approved_at 0.0016089942780140988
order_delivered_carrier_date 0.017930229985619613
order_delivered_customer_date 0.02981667521444877
order_estimated_delivery_date 0.0


In [11]:
# let's count the occurence of order_status values in rows where we have no missing values in any column
orders.dropna()['order_status'].value_counts()

order_status
delivered    96455
canceled         6
Name: count, dtype: int64

In [12]:
# in comparison, let's count the occurence of order_status values in rows that have at least one missing value in any column
orders[orders.isna().any(axis=1)]['order_status'].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered        23
created           5
approved          2
Name: count, dtype: int64

**What do we observe?** For the instances that have at least one missing value, we notice that the majority (i.e., >99%) of these orders were not delivered or impacted under various stages of the process, for instance, shipped, cancelled, unavailable and so on. So, if we are only interested in the **succesful deliveries** with no missing values, we can consider dropping these instances.

In [13]:
# we only look at orders that actually got delivered
orders = orders[orders['order_status'] == 'delivered']

In [14]:
# however, we still have to deal with missing values
for col in orders.columns:
    missings = len(orders[col][orders[col].isna()])
    print(col, missings)

order_id 0
customer_id 0
order_status 0
order_purchase_timestamp 0
order_approved_at 14
order_delivered_carrier_date 2
order_delivered_customer_date 8
order_estimated_delivery_date 0


In [15]:
# show the 23 problematic rows
missing_values_indexes = orders.index[orders.isna().any(axis=1)]
rows_with_missing_values = orders[orders.isna().any(axis=1)]
rows_with_missing_values

order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  e79f0f134f65def68b2af1eb9969f58c   
5323   e04abd8149ef81b95221e88f6ed9ab6a  212654a74f50461c222367b79cea5e3c   
16567  8a9adc69528e1001fc68dd0aaebbb54a  e76656ce4486a41da00e471277b1d1e9   
19031  7013bcfc1c97fe719a7b5e05e61c12db  919279990314ce9290fe3fd3bca96ee8   
20618  f5dd62b788049ad9fc0526e3ad11a097  2238c417bc745d8fa7d7389d8b22154b   
22663  5cf925b116421afa85ee25e99b4c34fb  54f6cbc1087187d1603082eb0594d47c   
23156  12a95a3c06dbaec84bcfb0e2da5d228a  beeda72b31be3b8a38b5c2b77d7705c4   
26800  c1d4211b3dae76144deccd6c74144a88  20b7cb7693554c07cc5cb8388f68d74c   
38290  d69e5d356402adc8cf17e08b5033acfb  86f9e3637836ff95e8e339223d2a136b   
39334  d77031d6a3c8a52f019764e68f211c69  ac9bcfbca17d0dee80c0d80125d55b71   
43834  2ebdfc4f15f23b91474edf87475f108e  f63db8b1515a5d3606d101d9b9cf9fe2   
48401  7002a78c79c519ac54022d4f8a65e6e8  6289cf6bd100be2499a029a27b1b5f26   
61743  2eecb0d85f281280f79fa00f9cec1a95  93d7e029a39295f3bc9b3c272365332e   
63052  51eb2eebd5d76a24625b31c33dd41449  01961d1b276640f460e6061e7f28873f   
67697  88083e8f64d95b932164187484d90212  491609eeefc72ee67ee5b42e63ba9859   
72407  3c0b8706b065f9919d0505d3b3343881  5cd3b70cb7cf755c4c87a1f9ebd19b0f   
73222  2aa91108853cecb43c84a5dc5b277475  53aeb29dc2e7dd0d2259f60ae35c157a   
79263  e69f75a717d64fc5ecdfae42b2e8e086  367605b57e5253221b9d9fd586e4bf89   
82868  0d3268bad9b086af767785e3f0fc0133  f57d49522c654b465c26e0cdf392fea6   
84999  2babbb4b15e6d2dfe95e2de765c97bce  a379939b4e880bfb2522dffc9ffbb54a   
92643  2d858f451373b04fb5c984a1cc2defaf  e9aea06ea89f977a316cdc731e0ab7db   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  3bab4e7cd9fd72391f090911718978d2   
98038  20edc82cf5400ce95e1afacc25798b31  6fc16fa677c8528424093b92165f7ddb   

      order_status order_purchase_timestamp   order_approved_at  \
3002     delivered      2017-11-28 17:44:07 2017-11-28 17:56:40   
5323     delivered      2017-02-18 14:40:00                 NaT   
16567    delivered      2017-02-18 12:45:31                 NaT   
19031    delivered      2017-02-18 13:29:47                 NaT   
20618    delivered      2018-06-20 06:58:43 2018-06-20 07:19:05   
22663    delivered      2017-02-18 16:48:35                 NaT   
23156    delivered      2017-02-17 13:05:55                 NaT   
26800    delivered      2017-01-19 12:48:08                 NaT   
38290    delivered      2017-02-19 01:28:47                 NaT   
39334    delivered      2017-02-18 11:04:19                 NaT   
43834    delivered      2018-07-01 17:05:11 2018-07-01 17:15:12   
48401    delivered      2017-01-19 22:26:59                 NaT   
61743    delivered      2017-02-17 17:21:55                 NaT   
63052    delivered      2017-02-18 15:52:27                 NaT   
67697    delivered      2017-02-18 22:49:19                 NaT   
72407    delivered      2017-02-17 15:53:27                 NaT   
73222    delivered      2017-09-29 08:52:58 2017-09-29 09:07:16   
79263    delivered      2018-07-01 22:05:55 2018-07-01 22:15:14   
82868    delivered      2018-07-01 21:14:02 2018-07-01 21:29:54   
84999    delivered      2017-02-18 17:15:03                 NaT   
92643    delivered      2017-05-25 23:22:43 2017-05-25 23:30:16   
97647    delivered      2018-06-08 12:09:39 2018-06-08 12:36:39   
98038    delivered      2018-06-27 16:09:12 2018-06-27 16:29:30   

      order_delivered_carrier_date order_delivered_customer_date  \
3002           2017-11-30 18:12:23                           NaT   
5323           2017-02-23 12:04:47           2017-03-01 13:25:33   
16567          2017-02-23 09:01:52           2017-03-02 10:05:06   
19031          2017-02-22 16:25:25           2017-03-01 08:07:38   
20618          2018-06-25 08:05:00                           NaT   
22663          2017-02-22 11:23:10           2017-03-09 07:28:47   
23156          2017-02-22 11:23:11           2017-03-02 11:09:19   
26800          2017-01-25 14:

Let's create some new variables to **impute these missing values**:
<ol>
    <li> <code>purchased_approved</code>: the seconds taken for an order to get approved after the customer purchases it</li>
    <li> <code>approved_carrier</code>: the hours taken for the order to go to the delivery carrier after it being approved</li>
    <li> <code>carrier_delivered</code>: the hours taken for the order to be delivered to the customer from the date it reaches the delivery carrier</li>
    <li> <code>delivered_estimated</code>: the hours difference between the estimated delivery date and the actual delivery date</li>
    <li> <code>purchased_delivered</code>: the hours taken for the order to be delivered to the customer from the date the customer made the purchase</li>
</ol>

In [16]:
# let's create the variables described above
orders['purchased_approved'] = (orders['order_approved_at'] - orders['order_purchase_timestamp']).dt.total_seconds() # in seconds
orders['approved_carrier'] = (orders['order_delivered_carrier_date'] - orders['order_approved_at']).dt.total_seconds() / 3600. # in hours
orders['carrier_delivered'] = (orders['order_delivered_customer_date'] - orders['order_delivered_carrier_date']).dt.total_seconds() / 3600. # in hours
orders['delivered_estimated'] = (orders['order_estimated_delivery_date'] - orders['order_delivered_customer_date']).dt.total_seconds() / 3600. # in hours
orders['purchased_delivered'] = (orders['order_delivered_customer_date'] - orders['order_purchase_timestamp']).dt.total_seconds() / 3600. # in hours

# show first 5 rows
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  a7c556ca59f5d0c765380ddafd09941d   
1  53cdb2fc8bc7dce0b6741e2150273451  9e285031fb5b57b3d1464d8672b8ed22   
2  47770eb9100c2d0c44946d9cf07ec65d  6348d8e73c3d01bb030637817b424016   
3  949d5b44dbf5de918fe9c16f97b45f8a  9b5d68f669f7ed215f789471556529fe   
4  ad21c59c0840e6cb83a9ceb5573f8159  42510cf6f88a1b5a38b09f09867c1f8d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  purchased_approved  approved_carrier  \
0                    2017-10-18               642.0         56.795833   
1                    2018-08-13            110570.0         11.109167   
2                    2018-09-04               994.0          4.910278   
3                    2017-12-15              1073.0         89.900000   
4                    2018-02-26              3710.0         21.434722   

   carrier_delivered  delivered_estimated  purchased_delivered  
0         145.503611           170.579722           202.477778  
1         288.945833           128.537500           330.768889  
2         220.274722           413.891944           225.461111  
3         226.811944           311.521667           317.010000  
4          46.507778           221.716111            68.973056

In [17]:
# get some statistics regarding the newly created (numeric) variables
orders.describe()

order_purchase_timestamp              order_approved_at  \
count                          96478                          96464   
mean   2018-01-01 23:29:31.939913984  2018-01-02 10:53:23.338012160   
min              2016-09-15 12:16:38            2016-09-15 12:16:38   
25%    2017-09-14 09:00:23.249999872            2017-09-14 14:48:54   
50%              2018-01-20 19:45:45            2018-01-22 13:50:04   
75%              2018-05-05 18:54:47  2018-05-06 10:57:24.750000128   
max              2018-08-29 15:00:37            2018-08-29 15:10:26   
std                              NaN                            NaN   

        order_delivered_carrier_date  order_delivered_customer_date  \
count                          96476                          96470   
mean   2018-01-05 05:01:50.912330240  2018-01-14 12:41:33.581683456   
min              2016-10-08 10:34:01            2016-10-11 13:46:32   
25%    2017-09-18 16:46:13.249999872     2017-09-25 22:15:09.500000   
50%       2018-01-24 16:13:47.500000            2018-02-02 19:32:21   
75%              2018-05-08 14:35:00     2018-05-15 22:54:48.500000   
max              2018-09-11 19:48:28            2018-10-17 13:22:46   
std                              NaN                            NaN   

       order_estimated_delivery_date  purchased_approved  approved_carrier  \
count                          96478        9.646400e+04      96462.000000   
mean   2018-01-25 17:09:52.325711616        3.699636e+04         67.177826   
min              2016-10-04 00:00:00        0.000000e+00      -4109.256111   
25%              2017-10-05 00:00:00        7.750000e+02         20.984722   
50%              2018-02-16 00:00:00        1.236000e+03         43.580278   
75%              2018-05-28 00:00:00        5.225250e+04         85.800208   
max              2018-10-25 00:00:00        2.669197e+06       3018.301667   
std                              NaN        7.392751e+04         84.889850   

       carrier_delivered  delivered_estimated  purchased_delivered  
count       96469.000000         96470.000000         96470.000000  
mean          223.927129           268.275017           301.397210  
min          -386.308056         -4535.401944            12.801944  
25%            98.401111           153.355556           162.388889  
50%           170.394444           286.754444           245.219444  
75%           288.687500           389.857569           377.284375  
max          4924.583333          3504.386944          5031.086667  
std           210.238168           244.424494           229.107753

Orders which have carrier date (i.e., `order_delivered_carrier_date`) prior to the date of order getting approved (i.e., `order_approved_at`), and orders which have delivered date (i.e. `order_delivered_customer_date`) prior to the carrier date (i.e. `order_delivered_carrier_date`) are assumed to be **corrupted data**, as it could not be logically true. Therefore, such records are **dropped**.

In [18]:
# drop rows where approved_carrier is less than 0, i.e. where the order was approved after the carrier picked it up
orders = orders.drop(orders[orders['approved_carrier'] < 0].index)
len(orders)

95128

In [19]:
# drop rows where carrier_delivered is less than 0, i.e. where the carrier delivered the order before picking it up
orders = orders.drop(orders[orders['carrier_delivered'] < 0].index)
len(orders)

95105

In [20]:
# check if we still have missing values
for col in orders.columns:
    missings = len(orders[col][orders[col].isna()])
    print(col, missings)

order_id 0
customer_id 0
order_status 0
order_purchase_timestamp 0
order_approved_at 14
order_delivered_carrier_date 2
order_delivered_customer_date 8
order_estimated_delivery_date 0
purchased_approved 14
approved_carrier 16
carrier_delivered 9
delivered_estimated 8
purchased_delivered 8


We will impute the missing values by **using the median** for each of our newly created variables

In [21]:
# calculate the median of the purchased_approved column
median_purchase_approved = orders['purchased_approved'].median()
print(median_purchase_approved) # in seconds

# replace missing values in order_approved_at by adding the calculated median to order_purchase_timestamp
orders['order_approved_at'].fillna(orders['order_purchase_timestamp'] + pd.Timedelta(seconds=median_purchase_approved), inplace=True)

1216.0


In [22]:
for col in orders.columns:
    missings = len(orders[col][orders[col].isnull()])
    print(col, missings)

order_id 0
customer_id 0
order_status 0
order_purchase_timestamp 0
order_approved_at 0
order_delivered_carrier_date 2
order_delivered_customer_date 8
order_estimated_delivery_date 0
purchased_approved 14
approved_carrier 16
carrier_delivered 9
delivered_estimated 8
purchased_delivered 8


In [23]:
# calculate the median of the approved_carrier column
median_approved_carrier = orders['approved_carrier'].median()
print(median_approved_carrier) # in hours

# replace missing values in order_delivered_carrier_date by adding the calculated median to order_approved_at
orders['order_delivered_carrier_date'].fillna(orders['order_approved_at'] + pd.Timedelta(hours=median_approved_carrier), inplace=True)

44.37277777777778


In [24]:
# check missing values
for col in orders.columns:
    missings = len(orders[col][orders[col].isnull()])
    print(col, missings)

order_id 0
customer_id 0
order_status 0
order_purchase_timestamp 0
order_approved_at 0
order_delivered_carrier_date 0
order_delivered_customer_date 8
order_estimated_delivery_date 0
purchased_approved 14
approved_carrier 16
carrier_delivered 9
delivered_estimated 8
purchased_delivered 8


In [25]:
# calculate the median of the carrier_delivered column
median_carrier_delivered = orders['carrier_delivered'].median()
print(median_carrier_delivered) # in hours

# replace missing values in order_delivered_customer_date by adding the calculated median to order_delivered_carrier_date
orders['order_delivered_customer_date'].fillna(orders['order_delivered_carrier_date'] + pd.Timedelta(hours=median_carrier_delivered), inplace=True)

170.64875


In [26]:
# we have to re-run our variable creation code from above
orders['purchased_approved'] = (orders['order_approved_at'] - orders['order_purchase_timestamp']).dt.total_seconds() # in seconds
orders['approved_carrier'] = (orders['order_delivered_carrier_date'] - orders['order_approved_at']).dt.total_seconds() / 3600. # in hours
orders['carrier_delivered'] = (orders['order_delivered_customer_date'] - orders['order_delivered_carrier_date']).dt.total_seconds() / 3600. # in hours
orders['delivered_estimated'] = (orders['order_estimated_delivery_date'] - orders['order_delivered_customer_date']).dt.total_seconds() / 3600. # in hours
orders['purchased_delivered'] = (orders['order_delivered_customer_date'] - orders['order_purchase_timestamp']).dt.total_seconds() / 3600. # in hours

In [27]:
# we shouldn't have any missing values anymore
for col in orders.columns:
    missings = len(orders[col][orders[col].isnull()]) / float(len(orders))
    print(col, missings)

order_id 0.0
customer_id 0.0
order_status 0.0
order_purchase_timestamp 0.0
order_approved_at 0.0
order_delivered_carrier_date 0.0
order_delivered_customer_date 0.0
order_estimated_delivery_date 0.0
purchased_approved 0.0
approved_carrier 0.0
carrier_delivered 0.0
delivered_estimated 0.0
purchased_delivered 0.0


In [28]:
# show imputed result and check whether they are chronological correct, this should be the case!
orders.loc[missing_values_indexes]

order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  e79f0f134f65def68b2af1eb9969f58c   
5323   e04abd8149ef81b95221e88f6ed9ab6a  212654a74f50461c222367b79cea5e3c   
16567  8a9adc69528e1001fc68dd0aaebbb54a  e76656ce4486a41da00e471277b1d1e9   
19031  7013bcfc1c97fe719a7b5e05e61c12db  919279990314ce9290fe3fd3bca96ee8   
20618  f5dd62b788049ad9fc0526e3ad11a097  2238c417bc745d8fa7d7389d8b22154b   
22663  5cf925b116421afa85ee25e99b4c34fb  54f6cbc1087187d1603082eb0594d47c   
23156  12a95a3c06dbaec84bcfb0e2da5d228a  beeda72b31be3b8a38b5c2b77d7705c4   
26800  c1d4211b3dae76144deccd6c74144a88  20b7cb7693554c07cc5cb8388f68d74c   
38290  d69e5d356402adc8cf17e08b5033acfb  86f9e3637836ff95e8e339223d2a136b   
39334  d77031d6a3c8a52f019764e68f211c69  ac9bcfbca17d0dee80c0d80125d55b71   
43834  2ebdfc4f15f23b91474edf87475f108e  f63db8b1515a5d3606d101d9b9cf9fe2   
48401  7002a78c79c519ac54022d4f8a65e6e8  6289cf6bd100be2499a029a27b1b5f26   
61743  2eecb0d85f281280f79fa00f9cec1a95  93d7e029a39295f3bc9b3c272365332e   
63052  51eb2eebd5d76a24625b31c33dd41449  01961d1b276640f460e6061e7f28873f   
67697  88083e8f64d95b932164187484d90212  491609eeefc72ee67ee5b42e63ba9859   
72407  3c0b8706b065f9919d0505d3b3343881  5cd3b70cb7cf755c4c87a1f9ebd19b0f   
73222  2aa91108853cecb43c84a5dc5b277475  53aeb29dc2e7dd0d2259f60ae35c157a   
79263  e69f75a717d64fc5ecdfae42b2e8e086  367605b57e5253221b9d9fd586e4bf89   
82868  0d3268bad9b086af767785e3f0fc0133  f57d49522c654b465c26e0cdf392fea6   
84999  2babbb4b15e6d2dfe95e2de765c97bce  a379939b4e880bfb2522dffc9ffbb54a   
92643  2d858f451373b04fb5c984a1cc2defaf  e9aea06ea89f977a316cdc731e0ab7db   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  3bab4e7cd9fd72391f090911718978d2   
98038  20edc82cf5400ce95e1afacc25798b31  6fc16fa677c8528424093b92165f7ddb   

      order_status order_purchase_timestamp   order_approved_at  \
3002     delivered      2017-11-28 17:44:07 2017-11-28 17:56:40   
5323     delivered      2017-02-18 14:40:00 2017-02-18 15:00:16   
16567    delivered      2017-02-18 12:45:31 2017-02-18 13:05:47   
19031    delivered      2017-02-18 13:29:47 2017-02-18 13:50:03   
20618    delivered      2018-06-20 06:58:43 2018-06-20 07:19:05   
22663    delivered      2017-02-18 16:48:35 2017-02-18 17:08:51   
23156    delivered      2017-02-17 13:05:55 2017-02-17 13:26:11   
26800    delivered      2017-01-19 12:48:08 2017-01-19 13:08:24   
38290    delivered      2017-02-19 01:28:47 2017-02-19 01:49:03   
39334    delivered      2017-02-18 11:04:19 2017-02-18 11:24:35   
43834    delivered      2018-07-01 17:05:11 2018-07-01 17:15:12   
48401    delivered      2017-01-19 22:26:59 2017-01-19 22:47:15   
61743    delivered      2017-02-17 17:21:55 2017-02-17 17:42:11   
63052    delivered      2017-02-18 15:52:27 2017-02-18 16:12:43   
67697    delivered      2017-02-18 22:49:19 2017-02-18 23:09:35   
72407    delivered      2017-02-17 15:53:27 2017-02-17 16:13:43   
73222    delivered      2017-09-29 08:52:58 2017-09-29 09:07:16   
79263    delivered      2018-07-01 22:05:55 2018-07-01 22:15:14   
82868    delivered      2018-07-01 21:14:02 2018-07-01 21:29:54   
84999    delivered      2017-02-18 17:15:03 2017-02-18 17:35:19   
92643    delivered      2017-05-25 23:22:43 2017-05-25 23:30:16   
97647    delivered      2018-06-08 12:09:39 2018-06-08 12:36:39   
98038    delivered      2018-06-27 16:09:12 2018-06-27 16:29:30   

      order_delivered_carrier_date order_delivered_customer_date  \
3002           2017-11-30 18:12:23       2017-12-07 20:51:18.500   
5323           2017-02-23 12:04:47       2017-03-01 13:25:33.000   
16567          2017-02-23 09:01:52       2017-03-02 10:05:06.000   
19031          2017-02-22 16:25:25       2017-03-01 08:07:38.000   
20618          2018-06-25 08:05:00       2018-07-02 10:43:55.500   
22663          2017-02-22 11:23:10       2017-03-09 07:28:47.000   
23156          2017-02-22 11:23:11       2017-03-02 11:09:19.000   
26800          2017-01-25 14:

In [29]:
# get descriptives on numerical variables
orders.describe()

order_purchase_timestamp              order_approved_at  \
count                          95105                          95105   
mean   2017-12-31 02:34:27.376015616  2017-12-31 12:13:21.022480640   
min              2016-09-15 12:16:38            2016-09-15 12:16:38   
25%              2017-09-12 21:39:28            2017-09-13 10:15:16   
50%              2018-01-18 01:16:32            2018-01-18 11:55:13   
75%              2018-05-03 18:54:48            2018-05-04 03:31:43   
max              2018-08-29 15:00:37            2018-08-29 15:10:26   
std                              NaN                            NaN   

        order_delivered_carrier_date  order_delivered_customer_date  \
count                          95105                          95105   
mean   2018-01-03 08:40:16.230955264  2018-01-12 17:24:46.494054144   
min              2016-10-08 10:34:01            2016-10-11 13:46:32   
25%              2017-09-15 15:17:29            2017-09-23 15:28:07   
50%              2018-01-22 21:28:40            2018-01-31 19:16:10   
75%              2018-05-07 14:19:00            2018-05-14 19:48:48   
max              2018-09-11 19:48:28            2018-10-17 13:22:46   
std                              NaN                            NaN   

       order_estimated_delivery_date  purchased_approved  approved_carrier  \
count                          95105        9.510500e+04      95105.000000   
mean   2018-01-23 20:29:15.920298496        3.473365e+04         68.448669   
min              2016-10-04 00:00:00        0.000000e+00          0.004167   
25%              2017-10-04 00:00:00        7.720000e+02         21.613333   
50%              2018-02-14 00:00:00        1.216000e+03         44.375278   
75%              2018-05-25 00:00:00        4.871000e+04         86.826111   
max              2018-10-25 00:00:00        2.669197e+06       3018.301667   
std                              NaN        6.881437e+04         83.584655   

       carrier_delivered  delivered_estimated  purchased_delivered  
count       95105.000000         95105.000000         95105.000000  
mean          224.741740           267.074841           302.838644  
min             0.000000         -4535.401944            12.801944  
25%            98.685000           152.761111           163.504444  
50%           170.648750           278.170833           246.521667  
75%           289.369167           389.451111           379.348333  
max          4924.583333          3504.386944          5031.086667  
std           210.467035           244.835579           229.714976

In [30]:
# get descriptives on non-numerical variables
orders.describe(include=object)

order_id                       customer_id  \
count                              95105                             95105   
unique                             95105                             61150   
top     e481f51cbdc54678b7cc49136f2d6af7  ae6a7336a0d9c66679da191b5bd246a6   
freq                                   1                                 8   

       order_status  
count         95105  
unique            1  
top       delivered  
freq          95105

In [31]:
# percentatge of orders that were delivered after the estimated delivery date
len(orders[orders['order_estimated_delivery_date'] <= orders['order_delivered_customer_date']]) / len(orders)

0.08194101256505967

In [32]:
# NOTE: this is the same as checking whether delivered_estimated is negative
len(orders[orders['order_estimated_delivery_date'] <= orders['order_delivered_customer_date']]) / len(orders) == len(orders[orders['delivered_estimated'] < 0]) / len(orders)

True

From the **summary statistics** of the succesfully delivered orders, we can infer that:

1. In 50% of all succesful deliveries it took less than 1216 seconds (or around 20 minutes) for the order to get approved after the customer purchased it.
2. On average, it took 303 hours (or approximately 12.6 days) for the order to get delivered to the customer from the date of purchase.
3. Since we only considered orders that actually got delivered, the order_status feature has only one class, i.e., delivered.
4. The first order was placed on 15/09/2016 and the last order was placed on 29/08/2018, from the available dataset.
5. On average, it took 267 hours (or approximately 11 days) less for an order to get delivered to the customer compared to the estimated date of delivery. However, around 8% of orders were delivered later than the estimated date of delivery.

## 2. Order Payments & Order Items

In [33]:
# read in the order_payments dataset
order_payments = pd.read_csv("./Data/order_payments.csv")

# show
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [34]:
# compare the number of unique order_ids in orders with the number of observations
print(len(order_payments['order_id'].unique()))
print(len(order_payments))

99440
103886


By comparing the unique order_ids with the number of observations in the dataframe, we can conclude that an `order_id` can have more than one payment!

In [35]:
# count the occurrences of each value in order_id
value_counts = order_payments['order_id'].value_counts()

# get the values with more than one occurrence for order_id
more_than_once = value_counts[value_counts > 1].index.tolist()

# subset based on the order_id
subset_df = order_payments[order_payments['order_id'].isin(more_than_once)].reset_index(drop=True)

# get the shape of the subset
subset_df.shape

(7407, 5)

In [36]:
# let's have a look at one specific order_id
print("order_id under investigation:", more_than_once[0])
specific_order_id = subset_df[subset_df['order_id'] == more_than_once[0]]
specific_order_id.sort_values('payment_sequential') # order based on sequence of payment

order_id under investigation: fa65dad1b0e818e3ccc5cb0e39231352


order_id  payment_sequential payment_type  \
1022  fa65dad1b0e818e3ccc5cb0e39231352                   1      voucher   
1642  fa65dad1b0e818e3ccc5cb0e39231352                   2   debit_card   
4732  fa65dad1b0e818e3ccc5cb0e39231352                   3      voucher   
723   fa65dad1b0e818e3ccc5cb0e39231352                   4      voucher   
1996  fa65dad1b0e818e3ccc5cb0e39231352                   5   debit_card   
2088  fa65dad1b0e818e3ccc5cb0e39231352                   6   debit_card   
5872  fa65dad1b0e818e3ccc5cb0e39231352                   7   debit_card   
4930  fa65dad1b0e818e3ccc5cb0e39231352                   8   debit_card   
1238  fa65dad1b0e818e3ccc5cb0e39231352                   9   debit_card   
1378  fa65dad1b0e818e3ccc5cb0e39231352                  10   debit_card   
2296  fa65dad1b0e818e3ccc5cb0e39231352                  11       paypal   
5263  fa65dad1b0e818e3ccc5cb0e39231352                  12   debit_card   
5565  fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
2615  fa65dad1b0e818e3ccc5cb0e39231352                  14   debit_card   
4548  fa65dad1b0e818e3ccc5cb0e39231352                  15   debit_card   
5576  fa65dad1b0e818e3ccc5cb0e39231352                  16      voucher   
5694  fa65dad1b0e818e3ccc5cb0e39231352                  17   debit_card   
5733  fa65dad1b0e818e3ccc5cb0e39231352                  18   debit_card   
5339  fa65dad1b0e818e3ccc5cb0e39231352                  19   debit_card   
6653  fa65dad1b0e818e3ccc5cb0e39231352                  20       paypal   
5309  fa65dad1b0e818e3ccc5cb0e39231352                  21   debit_card   
7163  fa65dad1b0e818e3ccc5cb0e39231352                  22      voucher   
6047  fa65dad1b0e818e3ccc5cb0e39231352                  23      voucher   
7068  fa65dad1b0e818e3ccc5cb0e39231352                  24       paypal   
1762  fa65dad1b0e818e3ccc5cb0e39231352                  25   debit_card   
5688  fa65dad1b0e818e3ccc5cb0e39231352                  26   debit_card   
357   fa65dad1b0e818e3ccc5cb0e39231352                  27   debit_card   
2781  fa65dad1b0e818e3ccc5cb0e39231352                  28   debit_card   
2780  fa65dad1b0e818e3ccc5cb0e39231352                  29   debit_card   

      payment_installments  payment_value  
1022                     1           3.71  
1642                     1           8.51  
4732                     1           2.95  
723                      1          29.16  
1996                     1           0.66  
2088                     1           5.02  
5872                     1           0.32  
4930                     1          26.02  
1238                     1           1.08  
1378                     1          12.86  
2296                     1           4.03  
5263                     1           3.60  
5565                     1           0.00  
2615                     1           0.00  
4548                     1          14.04  
5576                     1          12.86  
5694                     1           3.72  
5733                     1           0.73  
5339                     1           5.02  
6653                     1         150.00  
5309                     1           4.11  
7163                     1           4.03  
6047                     1          18.86  
7068                     1           0.42  
1762                     1           3.68  
5688                     1          28.27  
357                      1          66.02  
2781                     1          29.05  
2780                     1          19.26

In [37]:
# let's calculate the amount due
specific_order_id['payment_value'].sum()

457.99

The `order_payments` dataset is strongly related to the `order_items` dataset. More specifically, the `order_items` dataset contains a variable `price`. Let's compare whether our calculations correspond to the values found in order items.

In [38]:
# read in the order_items dataset
order_items = pd.read_csv("./Data/order_items.csv")

# show
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [39]:
# let's check the corresping row for our order_id under investigation
order_items[order_items['order_id'] == more_than_once[0]]

order_id  order_item_id  \
110202  fa65dad1b0e818e3ccc5cb0e39231352              1   

                              product_id                         seller_id  \
110202  1aecdb5fa3add74e385f25c6c527a462  06532f10282704ef4c69168b914b77be   

        shipping_limit_date   price  freight_value  
110202  2017-04-27 09:10:13  392.55          65.44

The values do not match for this specific instance! Since we know at least one instance for which this is the case, we will use the `order_payments` dataset (but you can aslo opt for the total sum of the price variable found in the `order_items` dataset) to get the total amount due by the customer.

In [40]:
# to do so, we will collapse the order_payments dataset so that we have one row for each unique order_id
# NOTE: this will drop the payment_sequential and payment_installments columns and the payment_type column will be the last one
grouped_df = order_payments.groupby(['order_id'])['payment_value'].sum().reset_index()

# rename column to 'total_amount_due'
grouped_df = grouped_df.rename(columns={'payment_value': 'total_amount_due'})

# show
grouped_df.head()

order_id  total_amount_due
0  00010242fe8c5a6d1ba2dd792cb16214             72.19
1  00018f77f2f0320c557190d7a144bdd3            259.83
2  000229ec398224ef6ca0657da4fc703e            216.87
3  00024acbcdf0a6daa1e931b038114c75             25.78
4  00042b26cf59d7ce69dfabb4e55b4fd9            218.04

In [41]:
# check to see if we find the same value as before
grouped_df[grouped_df['order_id'] == more_than_once[0]]

order_id  total_amount_due
97261  fa65dad1b0e818e3ccc5cb0e39231352            457.99

In Section 1 we mentioned to only investigate the succesful deliveries, so when we merge the two datasets, we only want to keep the `order_id`s that are in out pre-processed orders dataset.

In [42]:
# left merge with orders dataset (we add the payment value column)
orders_with_correct_price = pd.merge(orders, grouped_df, on='order_id', how='left')

# number of observations should be equal to the size of the orders dataset
orders_with_correct_price.shape

(95105, 14)

In [43]:
# show
orders_with_correct_price.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  a7c556ca59f5d0c765380ddafd09941d   
1  53cdb2fc8bc7dce0b6741e2150273451  9e285031fb5b57b3d1464d8672b8ed22   
2  47770eb9100c2d0c44946d9cf07ec65d  6348d8e73c3d01bb030637817b424016   
3  949d5b44dbf5de918fe9c16f97b45f8a  9b5d68f669f7ed215f789471556529fe   
4  ad21c59c0840e6cb83a9ceb5573f8159  42510cf6f88a1b5a38b09f09867c1f8d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  purchased_approved  approved_carrier  \
0                    2017-10-18               642.0         56.795833   
1                    2018-08-13            110570.0         11.109167   
2                    2018-09-04               994.0          4.910278   
3                    2017-12-15              1073.0         89.900000   
4                    2018-02-26              3710.0         21.434722   

   carrier_delivered  delivered_estimated  purchased_delivered  \
0         145.503611           170.579722           202.477778   
1         288.945833           128.537500           330.768889   
2         220.274722           413.891944           225.461111   
3         226.811944           311.521667           317.010000   
4          46.507778           221.716111            68.973056   

   total_amount_due  
0             38.71  
1            141.46  
2            179.12  
3             72.20  
4             28.62

When merging the `order_items` dataset it is important to know that an order can consist of multiple products that are sold by different sellers. You'll also notice the variable `shipping_limit_date`, which you can compare to the actual delivery date from the original orders dataset!

In [44]:
# drop price from order_items
order_items = order_items.drop('price', axis=1)

# left merge of order_items with orders_with_correct_price dataset (which includes the payment_value column)
order_items_payments = pd.merge(orders_with_correct_price, order_items, on='order_id', how='left')

# number of observations should be equal to the size of the orders dataset
order_items_payments.shape

(108605, 19)

In [45]:
# show
order_items_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  a7c556ca59f5d0c765380ddafd09941d   
1  53cdb2fc8bc7dce0b6741e2150273451  9e285031fb5b57b3d1464d8672b8ed22   
2  47770eb9100c2d0c44946d9cf07ec65d  6348d8e73c3d01bb030637817b424016   
3  949d5b44dbf5de918fe9c16f97b45f8a  9b5d68f669f7ed215f789471556529fe   
4  ad21c59c0840e6cb83a9ceb5573f8159  42510cf6f88a1b5a38b09f09867c1f8d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  purchased_approved  approved_carrier  \
0                    2017-10-18               642.0         56.795833   
1                    2018-08-13            110570.0         11.109167   
2                    2018-09-04               994.0          4.910278   
3                    2017-12-15              1073.0         89.900000   
4                    2018-02-26              3710.0         21.434722   

   carrier_delivered  delivered_estimated  purchased_delivered  \
0         145.503611           170.579722           202.477778   
1         288.945833           128.537500           330.768889   
2         220.274722           413.891944           225.461111   
3         226.811944           311.521667           317.010000   
4          46.507778           221.716111            68.973056   

   total_amount_due  order_item_id                        product_id  \
0             38.71              1  87285b34884572647811a353c7ac498a   
1            141.46              1  595fac2a385ac33a80bd5114aec74eb8   
2            179.12              1  aa4383b373c6aca5d8797843e5594415   
3             72.20              1  d0b61bfb1de832b15ba9d266ca96e5b0   
4             28.62              1  65266b2da20d04dbe00c5c2d3bb7859e   

                          seller_id  shipping_limit_date  freight_value  
0  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15           8.72  
1  289cdb325fb7e7f891c38608bf9e0962  2018-07-30 03:24:27          22.76  
2  4869f7a5dfa277a7dca6462dcf3b52b2  2018-08-13 08:55:23          19.22  
3  66922902710d126a0e7d26b0e3805106  2017-11-23 19:45:59          27.20  
4  2c9e548be18521d1c43cde1c582c6de8  2018-02-19 20:31:37           8.72

## 3. Products

In [46]:
# import products
products = pd.read_csv("./Data/products.csv")

# show
products.head()

product_id product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumery   
1  3aa071139cb16b67ca9e5dea641aaa2f                   art   
2  96bd76ec8810374ed1b65e291975717f        sports_leisure   
3  cef67bcfe19066a932b7673e239eb23d                  baby   
4  9dc1a7de274444849c219cff195d0b71            housewares   

   product_name_length  product_description_length  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [47]:
# check descriptives of numerical columns
products.describe()

product_name_length  product_description_length  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

**TIP**: you'll have to merge this dataset with the `order_items` dataset based on `product_id` if you want to use the details regarding the product. For example, you can calculate the volume of the product.

## 4. Customers

In [48]:
# import customers
customers = pd.read_csv("./Data/customers.csv")

# show
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   zip_code         city_name     province_name              region_name  \
0    1180.0             Ukkel               NaN  Brussels-Capital Region   
1    7760.0  Celles (Doornik)        Henegouwen           Walloon Region   
2    8951.0        Heuvelland   West-Vlaanderen           Flemish Region   
3    6870.0      Saint-Hubert         Luxemburg           Walloon Region   
4    8460.0         Oudenburg   West-Vlaanderen           Flemish Region   

  gender    birthday  
0      M  1979-05-15  
1      M  1976-04-29  
2      M  1988-12-23  
3      M  1991-09-22  
4      M  1970-08-23

In [49]:
# check data types
customers.dtypes

customer_id            object
customer_unique_id     object
zip_code              float64
city_name              object
province_name          object
region_name            object
gender                 object
birthday               object
dtype: object

**NOTE**: `zip_code` should be a string, not an integer or float. **Why?** Because it only has a select number of values and therefore does not have a continuous range!

In [50]:
# convert to string type
customers['zip_code'] = customers['zip_code'].astype(int).astype(str)

In [51]:
# check again
customers.dtypes

customer_id           object
customer_unique_id    object
zip_code              object
city_name             object
province_name         object
region_name           object
gender                object
birthday              object
dtype: object

In [52]:
# since we only have categorical features, we get count, unique, top, and the frequency of the most common value
customers.describe()

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       zip_code city_name province_name     region_name gender    birthday  
count     99441     99441         94927           99441  99441       99441  
unique     1142       587            10               3      2       26645  
top        1070      Pelt          Luik  Flemish Region      F  1940-06-03  
freq        286       398         13227           52470  51399          13

In [53]:
# check for missing values
customers.isnull().sum()

customer_id              0
customer_unique_id       0
zip_code                 0
city_name                0
province_name         4514
region_name              0
gender                   0
birthday                 0
dtype: int64

Missing values in province_name, but no missing values in zip-code. You can infer the province name easily. More information: https://www.spotzi.com/en/data-catalog/categories/postal-codes/belgium/

In [54]:
# function to extract province from zip code
def zip_code_province(zip_code):
    # initialize province
    province = None

    # retrieve first two characters of zip_code
    characters = zip_code[:2]

    # assign province
    if characters in ['10', '11', '12']:
        province = 'Brussel'
    elif characters in ['13', '14']:
        province = 'Waals Brabant'
    elif characters in ['15', '16', '17', '18', '19', '30', '31', '32', '33', '34']:
        province = 'Vlaams Brabant'
    elif characters in ['20', '21', '22', '23', '24', '25', '26', '27', '28', '29']:
        province = 'Antwerpen'
    elif characters in ['35', '36', '37', '38', '39']:
        province = 'Limburg'
    elif characters in ['40', '41', '42', '43', '44', '45', '46', '47', '48', '49']:
        province = 'Luik'
    elif characters in ['50', '51', '52', '53', '54', '55', '56', '57', '58', '59']:
        province = 'Namen'
    elif characters in ['60', '61', '62', '63', '64', '65', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79']:
        province = 'Henegouwen'
    elif characters in ['66', '67', '68', '69']:
        province = 'Luxemburg'
    elif characters in ['80', '81', '82', '83', '84', '85', '86', '87', '88', '89']:
        province = 'West-Vlaanderen'
    elif characters in ['90', '91', '92', '93', '94', '95', '96', '97', '98', '99']:
        province = 'Oost-Vlaanderen'

    return province

In [55]:
# apply function on dataframe
customers['province_name'] = customers['zip_code'].apply(lambda x: zip_code_province(x))

# check for missing values again
customers.isnull().sum()

customer_id           0
customer_unique_id    0
zip_code              0
city_name             0
province_name         0
region_name           0
gender                0
birthday              0
dtype: int64

In [56]:
# group by province and count number of customers
province_cust = customers.groupby('province_name')['customer_id'].count().sort_values(ascending=False).reset_index()

# rename columns
province_cust.rename(columns={'province_name': 'NE_Name', 'customer_id':'count'}, inplace=True)

# show
province_cust

NE_Name  count
0              Luik  13227
1         Antwerpen  12669
2        Henegouwen  12596
3    Vlaams Brabant  10998
4   Oost-Vlaanderen  10578
5   West-Vlaanderen  10272
6           Limburg   7953
7         Luxemburg   6122
8             Namen   5825
9     Waals Brabant   4687
10          Brussel   4514

**TIP**: You can use the `birthday` to calculate the age of a customer. Be careful,some birthdays can be incorrect. For example, they are too young (-18) to create an account or maybe too old to be feasible. 
You can either remove these observations or impute them with a proxy.

### Group Assignment

Although we haven't covered all the datasets in this notebook, it already helps you to get started with the group assignment. Moreover, based on this week's course material you are already able to solve some of required subquestions found on Ufora.